### PR TITLE
fixes #19 Michaelconnor/burn payto fix

### LINKED ISSUE #19: Possible front-running attack in burn functions
I’ve read in your security policy that you’d like to discuss potential vulnerabilities in issues, so I’m starting the discussion here.

In the burn function, there’s a proof that the caller is allowed to burn. Additionally, in the transaction submitted to the blockchain there’s a “payTo” address.

It seems that the “payTo” address is not bound to the proof in any way - e.g., by including it as a public input or signing it with a key bound to the proof.

Is there anything preventing a node or a miner from just copying the proof and submit a new transaction with another “payTo” address?

### PR DESCRIPTION
**Pull request by iAmMichaelConnor**
fixes #19

See README: added a 'Security Updates' section. When testing, please follow the steps in this section, to ensure they're correct.

Let's **tag** important security updates. (Let's also tag the first commit as v1.0.0).

See code changes to address the payTo bug.

### CODE REVIEW

**Comment by iAmMichaelConnor** [comment]
I should add: In order to link 'payTo' inextricably to the 'proof', I add a really simple logic gate to the .code file:
I check payTo * 1 == payTo.

**Comment by iAmMichaelConnor** [comment]
Negative test completed:
Generated a proof with Alice's payTo address.
When sending proof & public inputs to the smart contract, substituted a 'Thief's' address as the 'payTo' public input.
The transaction failed.

Console evidence below:

```zkp_1               |   Existing Proof Variables:
zkp_1               |     A: 0x12e18571b8c05600000000000000000000000000000000000000000000000000 : 25097077287736233400735450050399830016,0
zkp_1               |     sk_A: 0xdb01597340b274031779a206a53760e23c8545af471250c3c7f8fe : 264761126927143143264975266,8833220164816321136378610184173910270
zkp_1               |     Pk_A: 0xa83b272c2f0ff425ffd97962357c672e24263156ca249e8990745f : 203378879924546887948949881,130542058513995705326603209572913017951
zkp_1               |     S_A: 0x535df85d3086a758df11ec29ea0037f5d0504611a2fe4543420f0a : 100784604618466476192371180,55713349725703301797241750312951353098
zkp_1               |     z_A: 0x90b0ceb0e49887f825e8655480a938f9d21663347d9a8d524b6f88 : 174920267301251397999519845,112323197885141256781182878780039589768
zkp_1               |     payTo: 0xd0c5e65f4b73b2c032ac99dd0dbecb448ff47e14 : 3502630495,100292839318999760712988558633272114708
zkp_1               |   New Proof Variables:
zkp_1               |     Na: 0x66ed2306c76753aff62cbf8652b82b70cfe688a56888e83bee0c8d : 124430280581647555295063231,178546055182670300264992226517281803405
zkp_1               |     root: 0x31df2ad0fb1abfe229112958a0a2ff0bed19fb7d5e1a3022c8b6bf : 60291242708707046723227945,117806137083666375169726650334963087039
zkp_1               |   thiefInputs:
zkp_1               |   [ '320263376',
zkp_1               |     '175214591508463953737031282311922365229',
zkp_1               |     '25097077287736233400735450050399830016',
zkp_1               |     '0',
zkp_1               |     '124430280581647555295063231',
zkp_1               |     '178546055182670300264992226517281803405',
zkp_1               |     '60291242708707046723227945',
zkp_1               |     '117806137083666375169726650334963087039' ]
zkp_1               |   inputs:
zkp_1               |   [ '3502630495',
zkp_1               |     '100292839318999760712988558633272114708',
zkp_1               |     '25097077287736233400735450050399830016',
zkp_1               |     '0',
zkp_1               |     '124430280581647555295063231',
zkp_1               |     '178546055182670300264992226517281803405',
zkp_1               |     '60291242708707046723227945',
zkp_1               |     '117806137083666375169726650334963087039' ]
zkp_1               |   /app
zkp_1               |   gm17/nft-burn/
zkp_1               |   Computing proof with w=[sk_A,S_A,path[],order] x=[A,Na,root,1]
zkp_1               |     ZoKrates running with Nodejs environment docker
zkp_1               |     Running the container; mounted: nightfall_zkp-code:/home/zokrates/code/:cached
zkp_1               |     Container id: 54af0a3e0ed1b72730fc831481475d624ab75d73ab8774f177bacd0ece18f6b0
zkp_1               |     To connect to the container manually: 'docker exec -ti 54af0a3e0ed1b72730fc831481475d624ab75d73ab8774f177bacd0ece18f6b0 bash'
zkp_1               |     
zkp_1               |     Compute-witness: Executing the program C(w,x) with:
zkp_1               |      (w,x)= [ '3502630495',
zkp_1               |       '100292839318999760712988558633272114708',
zkp_1               |       '25097077287736233400735450050399830016',
zkp_1               |       '0',
zkp_1               |       '264761126927143143264975266',
zkp_1               |       '8833220164816321136378610184173910270',
zkp_1               |       '100784604618466476192371180',
zkp_1               |       '55713349725703301797241750312951353098',
zkp_1               |       '0',
zkp_1               |       '0',
zkp_1               |       '0',
zkp_1               |       '0',
zkp_1               |       '0',
zkp_1               |       '0',
zkp_1               |       '0',
zkp_1               |       '0',
zkp_1               |       '0',
zkp_1               |       '0',
zkp_1               |       '0',
zkp_1               |       '0',
zkp_1               |       '0',
zkp_1               |       '0',
zkp_1               |       '0',
zkp_1               |       '0',
zkp_1               |       '0',
zkp_1               |       '0',
zkp_1               |       '0',
zkp_1               |       '0',
zkp_1               |       '0',
zkp_1               |       '0',
zkp_1               |       '0',
zkp_1               |       '0',
zkp_1               |       '0',
zkp_1               |       '0',
zkp_1               |       '0',
zkp_1               |       '0',
zkp_1               |       '0',
zkp_1               |       '0',
zkp_1               |       '0',
zkp_1               |       '0',
zkp_1               |       '0',
zkp_1               |       '0',
zkp_1               |       '0',
zkp_1               |       '0',
zkp_1               |       '0',
zkp_1               |       '0',
zkp_1               |       '0',
zkp_1               |       '0',
zkp_1               |       '0',
zkp_1               |       '0',
zkp_1               |       '0',
zkp_1               |       '0',
zkp_1               |       '0',
zkp_1               |       '0',
zkp_1               |       '0',
zkp_1               |       '0',
zkp_1               |       '0',
zkp_1               |       '0',
zkp_1               |       '0',
zkp_1               |       '0',
zkp_1               |       '0',
zkp_1               |       '0',
zkp_1               |       '0',
zkp_1               |       '0',
zkp_1               |       '0',
zkp_1               |       '0',
zkp_1               |       '0',
zkp_1               |       '0',
zkp_1               |       '0',
zkp_1               |       '0',
zkp_1               |       '0',
zkp_1               |       '0',
zkp_1               |       '4294967295',
zkp_1               |       '124430280581647555295063231',
zkp_1               |       '178546055182670300264992226517281803405',
zkp_1               |       '60291242708707046723227945',
zkp_1               |       '117806137083666375169726650334963087039' ] ...
zkp_1               |     a  3502630495 100292839318999760712988558633272114708 25097077287736233400735450050399830016 0 264761126927143143264975266 8833220164816321136378610184173910270 100784604618466476192371180 55713349725703301797241750312951353098 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 4294967295 124430280581647555295063231 178546055182670300264992226517281803405 60291242708707046723227945 117806137083666375169726650334963087039
zkp_1               |     Writing new time estimate of 36650 to stats.json file...
zkp_1               |     
zkp_1               |     Generating Proof := P(pk,w,x)
zkp_1               |     Backend being used gm17
zkp_1               |     
zkp_1               |     Extracting the Proof from the container's console
zkp_1               |     Writing new time estimate of 125272 to stats.json file...
zkp_1               |     Proof: {
zkp_1               |       "A": [
zkp_1               |         "0x2719bd488cb036b40a003256e4df316171dcbbad77fefc4b2f573d1c98b4c46e",
zkp_1               |         "0x3b9d60f40b2406a4c9187b247f54ae2b60b0f68a3d63ed6da6cb939fc3aade1"
zkp_1               |       ],
zkp_1               |       "B": [
zkp_1               |         [
zkp_1               |           "0xa94f9956075c645e6c0b9ecd46ee42e954062dda52044d761fc5f0a4dbfb764",
zkp_1               |           "0x2c9959e5ea3e5952f5d58bc1936ff8d84083fbd0afacd568c7a57029f626e7ec"
zkp_1               |         ],
zkp_1               |         [
zkp_1               |           "0xb8592df7ccea5bb59372aa818e47da8adac41da86b7c6c2932de5864eb40a5",
zkp_1               |           "0x136acac12d67616ce5a4a3a4b5e091408a8324866c4846beeec7d1fd6f630488"
zkp_1               |         ]
zkp_1               |       ],
zkp_1               |       "C": [
zkp_1               |         "0x47e88d3e21c284c4d22a0c26625c3f13c90eed4bbf53e50f885b2f2adb360c8",
zkp_1               |         "0xa7eeab65cfd5191cbb05e8033807a0c9fecd7e3e48928611b6679e21326f74c"
zkp_1               |       ]
zkp_1               |     }
zkp_1               |     Killing the container
zkp_1               |   Burning within the Shield contract
zkp_1               |     proof:
zkp_1               |     [ '17685678657358763145099174749214955451222870992805671498899539948007668892782',
zkp_1               |       '1685282637664528584876514208462904608076453726757390240813842975008759655905',
zkp_1               |       '4786344413450698641333931612817042875062749845028929268850900793101301823332',
zkp_1               |       '20172713392496190137843080126876523608799667169375102381582883412198715484140',
zkp_1               |       '325715354638120472327050130660215646033466909456568649426849188009082568869',
zkp_1               |       '8782629272761909995971738738211867098092383882548462376982974241221495686280',
zkp_1               |       '2032818474359213469948906019007785159077202317169083832161578546926447714504',
zkp_1               |       '4747371141143122796517041841543353553140183071016318723825430846644901312332' ]
zkp_1               |     inputs:
zkp_1               |     [ '320263376',
zkp_1               |       '175214591508463953737031282311922365229',
zkp_1               |       '25097077287736233400735450050399830016',
zkp_1               |       '0',
zkp_1               |       '124430280581647555295063231',
zkp_1               |       '178546055182670300264992226517281803405',
zkp_1               |       '60291242708707046723227945',
zkp_1               |       '117806137083666375169726650334963087039',
zkp_1               |       '1' ]
zkp_1               |     vkId: 0x8ce4ffc7d60633ccd635fd92d709d9c4fc9f619d1df22c3338e96d8cfe063000
ganache_1           | eth_getBlockByNumber
ganache_1           | eth_sendTransaction
ganache_1           | 
ganache_1           |   Transaction: 0x25ffd05f9ea8d5d29790697a3747e2ed0340d92fe609411f07cf1a5457ed4211
ganache_1           |   Gas usage: 1957031
ganache_1           |   Block Number: 30
ganache_1           |   Block Time: Wed Jun 05 2019 15:04:42 GMT+0000 (Coordinated Universal Time)
ganache_1           |   Runtime Error: revert
ganache_1           |   Revert reason: The proof has not been verified by the contract
ganache_1           | 
ganache_1           | eth_call
zkp_1               |     { Error: Returned error: VM Exception while processing transaction: revert The proof has not been verified by the contract -- Reason given: The proof has not been verified by the contract.
zkp_1               |         at Object.ErrorResponse (/app/node_modules/web3-core-helpers/src/errors.js:29:16)
zkp_1               |         at /app/node_modules/web3-core-requestmanager/src/index.js:140:36
zkp_1               |         at XMLHttpRequest.request.onreadystatechange (/app/node_modules/web3/node_modules/web3-providers-http/src/index.js:91:13)
zkp_1               |         at XMLHttpRequestEventTarget.dispatchEvent (/app/node_modules/xhr2-cookies/xml-http-request-event-target.ts:44:13)
zkp_1               |         at XMLHttpRequest._setReadyState (/app/node_modules/xhr2-cookies/xml-http-request.ts:219:8)
zkp_1               |         at XMLHttpRequest._onHttpResponseEnd (/app/node_modules/xhr2-cookies/xml-http-request.ts:345:8)
zkp_1               |         at IncomingMessage.<anonymous> (/app/node_modules/xhr2-cookies/xml-http-request.ts:311:39)
zkp_1               |         at IncomingMessage.emit (events.js:198:15)
zkp_1               |         at endReadableNT (_stream_readable.js:1139:12)
zkp_1               |         at processTicksAndRejections (internal/process/task_queues.js:81:17) reason: 'The proof has not been verified by the contract' }
```

This gives a clear Solidity error (as expected and desired).

File: README.md [hunk 1]
```diff
@@ -24,6 +24,38 @@ As well as this file, please be sure to check out:
...
+
+**date**: 2019-06-04  
+**description**: Burn payTo bug  
+**tag \#**:  v1.0.1  
+**PR \#**: [#26](https://github.com/EYBlockchain/nightfall/pull/26)  
+**issue \#**: [#19](https://github.com/EYBlockchain/nightfall/issues/19)  
+**re-installation**:  
+`cd zkp-utils`  
+`npm ci`  
+`cd zkp`
```
Comment (asishap) [inline README.md]: Please add one more line before change to `zkp` directory
`cd ..`

Comment (iAmMichaelConnor) [reply README.md]: Corrected, thanks!

**Pull request merged**